### DISCUSSION TITLE
Why `layout(..)` still exists?

### DISCUSSION THREAD

**Discussion by memeplex** [typst/typst#4732] [Q&A] [answered]
Today I was reading the Layout reference and was surprised that `layout` still exists as a non-deprecated separate function inside which one can measure things.

It's mentioned in [Laurenz's post](https://laurmaedje.github.io/posts/types-and-context/) as one of the three contextual functions:

> As it stands, locate, style, and layout are three separate instances of the same pattern

Also the documentation of `measure` is ambiguous, if not plainly contradictory:

> If you want to measure in the current layout dimensions, you can combine measure and layout.

> For this reason, you can only measure when context is available.

Is this "current container" context not actually a context? What's the rationale behind this?

**Comment by memeplex** [comment]
To be more precise, I would have expected for `layout` to still exist but with a caveat like the one for `locate`:

> Compatibility In Typst 0.10 and lower, the locate function took a closure that made the current location in the document available

OTOH `styles` was deprecated altogether which makes sense since the value passed to the callback was opaque and only observed through functions that are now contextual.

**Comment by laurmaedje** [comment, marked as answer]
The problem is that `layout` is only able to provide this information by forcing its contents to be a separate block (e.g. `a #layout(_=> [hi]) b` is three separate blocks). But we don't always want that for `context`. `a #context [hi] b` is a single paragraph.

`layout(_ => ..)` also provides context to `..` though. So essentially it can provide even more information than `context` at the cost of forcing a separate block.

**Reply by memeplex** [reply]
I've been pondering on your answer. As I now understand it there are two parts to the context mechanism:

1. A number of functions that were contextual now implicitly take their input from the context, so you don't have to fiddle with opaque objects that were intended to be passed to those functions.
2. A new syntax that even makes the callback implicit. Instead of `contextual(() => ...)` you just write `#context {...}`.

Perhaps what confused me was that I expected, given that it got its own syntax, and specially after reading your post, for `#context {...}` to subsume all cases that depended on context. 

So the way to understand `layout` may be like a function that exposes context but also does something more, as if it were defined somewhat like `layout(cb) = { ... context { ... cb(...) ... } ... }`. This way `#context { ... }` is still the only primitive able to expose context. What I like about this way of thinking is that we have two building blocks: contexts and functions (that may open contexts as part of their operation). And what I'm trying to avoid is an alternative view in which you have contexts but also functions that somehow provide contexts that are not quite contexts or proper contexts, or maybe are extended contexts, etc, on a case-by-case basis. Is my description accurate at a conceptual level?

**Reply by laurmaedje** [reply]
Yes, 1. + 2. are accurate.

>  This way #context { ... } is still the only primitive able to expose context. What I like about this way of thinking is that we have two building blocks: contexts and functions (that may open contexts as part of their operation).

I think this way of thinking mostly works, but it's not fully true to the implementation. During styling & layout, the contextual data is simply available. All `context` does it delay a computation until realization. Show rules run during realization, so they also provide context. Meanwhile, `layout` delays a computation even further, until layout. There it can provide everything `context` provides, plus the available size.

> And what I'm trying to avoid is an alternative view in which you have contexts but also functions that somehow provide contexts that are not quite contexts or proper contexts, or maybe are extended contexts.

The context provided by `layout` (and also by show rules) is a normal "proper" context. It doesn't internally need to invoke `context` because the information is already available, but it construct the same internal `Context` struct that is made available to code.

I think your concerns are somewhat valid still. Ideally, I'd also like for the handling of layout and the other contextual things to be more uniform. Perhaps, we manage to do that in the future.

**Reply by memeplex** [reply]
> the handling of layout and the other contextual things to be more uniform

Perhaps it may be something like `context [stage]` with a default of `realize` but that can be set to `layout`, with `container-size` as a exposed function (or attribute, or some other api) and the different behavior regarding blocks. This is not really more convenient, but I believe it would more clearly convey the ideas you expressed here. set rules are already a different kind of beast, so it's ok.

**Reply by memeplex** [reply]
> Show rules run during realization, so they also provide context.

One remark about this. I realized that page show rules don't provide context. Perhaps this is a deliberate exception, but I can report it if not.

**Reply by laurmaedje** [reply]
Do you mean `show page`? These are broken to the point that I don't really see a good use for them currently. I have plans for making them useful in the future.

Generally, it's also the case that all show rules (except for bare `show:` which is a bit the odd one out right now) provide [style context](https://typst.app/docs/reference/context/#style-context), but only some provide [location context](https://typst.app/docs/reference/context/#location-context).
